Introduce a Remote Tagger Client

juliogreff opened the pull request:
### What does this PR do?

Back in #6380 we introduced the server-side of this feature. This commit introduces the client-side
part, by moving the previous tagger to pkg/tagger/local.Tagger, and
introducing the new client of that in pkg/tagger/remote.Tagger. Both
satisfy the same pkg/tagger.Tagger interface, so switching between the
two is mostly transparent.

### Describe your test plan

This PR cannot be tested on its own, as it does not contain any entry points to use the new feature. Entry points have been added in #7154, so you can check that PR for instructions on how to test this.

juliogreff committed:
Introduce a gRPC client for the Remote Tagger

Back in 666c775b3b0b408bc8ba4593f420fd526cd841a1 we introduced the
server-side of this feature. This commit introduces the client-side
part, by moving the previous tagger to pkg/tagger/local.Tagger, and
introducing the new client of that in pkg/tagger/remote.Tagger. Both
satisfy the same pkg/tagger.Tagger interface, so switching between the
two is mostly transparent.

juliogreff committed:
Log when remote tagger starts

vboulineau commented on pkg/tagger/remote/tagger.go:
```diff
@@ -0,0 +1,270 @@
+		// the new stream is responsible for re-adding all the still
+		// existing entities back. for the time in between, this can
+		// cause queries to the tagger to return nothing.
+		t.store.reset()
```
Should we add a flag (internal) to flag remote readiness? In the spirit of `informers.WaitForCacheSync`, this would allow to log in `Tag()` requests received while tagger is not ready. It could be extremely useful in flares we get on support tickets.

juliogreff replied:
Good point. I think I'll have to change the gRPC protos to stream a list of events, instead of event by event, to be able to make this work. In that case, the first batch of events would mean the tagger is ready (otherwise, with event-by-event, we wouldn't know). That way I think we can also avoid the no tags case entirely, by keeping the old state of the store until we get a new batch. Changes incoming ;)

vboulineau commented on pkg/tagger/remote/tagger.go:
```diff
@@ -0,0 +1,270 @@
+
+	t.client = pb.NewAgentSecureClient(t.conn)
+
+	err = t.startTaggerStream(5 * time.Minute)
```
const it for readability

juliogreff commented on pkg/tagger/remote/tagger.go:
```diff
@@ -0,0 +1,270 @@
+		fmt.Sprintf(":%v", config.Datadog.GetInt("cmd_port")),
+		grpc.WithTransportCredentials(creds),
+	)
+	if err != nil {
```
Yeah, `startTaggerStream` usually fails at least once when starting the agent, and it gets retried properly 👍

juliogreff committed:
Atomically replace remote tagger contents

This solves an issue when a tagger stream is aborted, and for a small
amount of time the tag store is empty until it is rebuilt event by
event. Now, it is rebuilt all at once in an event batch, and this new
state replaces the old one atomically, preventing empty tags entirely.

truthbk reviewed (approved):
grpc bits look 👍 , also took a superficial look at the remainder of the PR which also appears 👌 

Nice stuff!
